File: 01.Python/0977.py
from typing import List

class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [0] * n

        left = 0
        right = n - 1
        write = n - 1

        while left <= right:
            if abs(nums[left]) > abs(nums[right]):
                result[write] = nums[left] * nums[left]
                left += 1
            else:
                result[write] = nums[right] * nums[right]
                right -= 1
            
            write -= 1
        
        return result

sol = Solution()
nums = [-7, -3, 2, 3, 11]
print(sol.sortedSquares(nums))

File: 03.CPP/0977.cpp
#include <iostream>
#include <vector>
#include <cmath>

class Solution {
public:
    std::vector<int> sortedSquares(std::vector<int>& nums) {
        int n = nums.size();
        std::vector<int> result(n);

        int left = 0;
        int right = n - 1;
        int write = n - 1;

        while (left <= right) {
            if (abs(nums[left]) > abs(nums[right])) {
                result[write] = nums[left] * nums[left];
                left++;
            } else {
                result[write] = nums[right] * nums[right];
                right--;
            }

            write--;
        }

        return result;
    }
};

void printVector(const std::vector<int>& vec) {
    std::cout << "[";
    for (int i = 0; i < vec.size(); i++) {
        std::cout << vec[i] << (i < vec.size() - 1 ? ", " : "");
    }
    std::cout << "]" << std::endl;
}

int main() {
    Solution sol;

    std::vector<int> nums = {-7, -3, 2, 3, 11};
    std::vector<int> result = sol.sortedSquares(nums);

    printVector(result);
    return 0;
}